### ISSUE TITLE
cudnn v6 support

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#5734]
Since issue https://github.com/Theano/Theano/issues/5679, we make Theano compile with it. We should do:

- [ ] Make a todo list of all the new features we need to wrap, to help do the follow up.
- [ ] Test that our warning in https://github.com/Theano/Theano/pull/5427 isn't useful anymore and remove it.

**Comment by Dapid** [comment]
The new features are:

- [ ] Dilated convolutions
- [ ] Convolution, bias, and activation fused in a single kernel. (Only ReLU and per channel bias).
- [ ] Inference on int8 dtype.
- [ ] 2 new implementations of RNN (Pascal only)
- [ ] 1D FFT convolution
- [ ] `cudnnReduceTensor`, supporting 8 reduce operations
- [ ] ELU activation
- [ ] Deterministic max pooling

**Comment by nouiz** [comment]
Thanks, I put that in the description.

**Comment by Dapid** [comment]
I am doing a lot of 1D convolutions, so I'd would like to help with that. Any pointers on where to start?

Also, regarding the ELU, it will probably not make a big difference, but it seems easy enough.

**Comment by HapeMask** [comment]
> Convolution, bias, and activation fused in a single kernel. (Only ReLU and per channel bias).
>> Is it worth it?

I've seen at least one paper mention eliminating biases to avoid extra kernel calls where possible: https://arxiv.org/pdf/1606.02147.pdf (but they don't say how much performance is gained by reducing kernel calls).

I took a quick look to see if it could be easily made into a local optimization, but since ReLU doesn't currently have its own Op I'm not sure how to reliably detect where this optimization could be applied.

**Comment by nouiz** [comment]
I think the speed up for big model is small. Also, that optimization maybe
could be applied only for the inference depending how it was implemented.
Any idea is we can use it for gradient?

It won't request just an optimizer, but also c code changes.

**Comment by HapeMask** [comment]
> maybe could be applied only for the inference

You're right, I just looked at the cudnn docs and v6 only adds ```cudnnConvolutionBiasActivationForward``` so it probably wouldn't help training performance much.

**Comment by astooke** [comment]
Word to the wise...when upgrading CuDNN version, make sure to call `theano-cache clear`.

Maybe this is obvious but I just lost a couple hours chasing an `ImportError` looking for libcudnn.so.5 despite the device initializing with CuDNN 6020.  Even when I made a new test file, using the same `theano.gpuarray.dnn.dnn_conv()` got me...no step along the traceback was human legible. (And I looked at this page in the meantime so thought I'd write here).

Working now :)
